### ISSUE TITLE
Mapped array/tuple types not always working as expected

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#43129]
# Bug Report

### 🔎 Search Terms

Mapped array types

### 🕗 Version & Regression Information

I noticed now in TS 4.1.2.

### ⏯ Code and Playground Links

There are two problems.

The following has an error in the mapped type despite that I did it just like in https://github.com/microsoft/TypeScript/pull/26063 (as far as I know):

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

type ElementTypeArrayToInstArray<T extends Constructor[]> = {
	[K in keyof T]: InstanceType<T[K]> // <----------- ERROR
}
type ArrayValues<T extends any[]> = T[keyof T]
type ElementTypes<T extends Constructor[]> = ArrayValues<ElementTypeArrayToInstArray<T>>

type test = ElementTypes<[typeof HTMLAnchorElement, typeof HTMLDivElement]>
```

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoV6EhQAogBsIALYcwATgEAyoqLwgBHAAksjAEVEghMxsHNywiCgY2LgimqTKrgCQggDSUPIIUADWECBwAGZQBMIWCSj8mBChkISVwpoA9CNQeAC00zOzkz4ASgsA8guuqh5hjJHRAGq8vugQSGms7Fw8-ELDYgSCDU2t7ZtefoHB-cenGRfZiXk4VCFMTJPYHI4nN5BBAhMKgmLxRLwwjqdTuTzQdgoMRQj5hE6CDEtKAACQIAFkADIMBCYAAWuFxMPoRNaZKpABF5AA3JnAYauIA)

But notice that when you hover on `test`, it is a union of all the values of the array including values from non-numeric keys, which is unlike the examples in #26063.

There seems to some sort of special casing happening.

I had to specify that I want `number` keys only in order(unlike #26063) to get rid of the error:

```js
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

type ElementTypeArrayToInstArray<T extends Constructor[]> = {
	[K in keyof T]: InstanceType<T[number & K]> // <----------- GOOD
}
type ArrayValues<T extends any[]> = T[keyof T]
type ElementTypes<T extends Constructor[]> = ArrayValues<ElementTypeArrayToInstArray<T>>

type test = ElementTypes<[typeof HTMLAnchorElement, typeof HTMLDivElement]>
```

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoV6EhQAogBsIALYcwATgEAyoqLwgBHAAksjAEVEghMxsHNywiCgY2LgimqTKrgCQggDSUPIIUADWECBwAGZQBMIWCSj8mBChkISCCOgB5BCoUI4VwpoA9LNQeAC0K6trS1AA4gDy2wAirqoeYYyR0QBqvL7oEEhprOxcPPxCM2IEgg1Nre3HXn6BYL9W73DJPbKJPI4VCFMTJC5XG53AFBBAhMLwmLxRKYwjqdTuTzQdgoMQooFhO6CIktKAACQIAFkADIMBCYAAWuHJaPoNNaDJZe3kADcecAZq4gA)

But notice that `test` still contains all the values of all keys, not just array values.

Finally, I had to specify numeric keys for `ArrayValues`:

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

type ElementTypeArrayToInstArray<T extends Constructor[]> = {
	[K in keyof T]: InstanceType<T[number & K]>
}
type ArrayValues<T extends any[]> = T[number & keyof T] // <-------- number
type ElementTypes<T extends Constructor[]> = ArrayValues<ElementTypeArrayToInstArray<T>>

type test = ElementTypes<[typeof HTMLAnchorElement, typeof HTMLDivElement]>
```

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoV6EhQAogBsIALYcwATgEAyoqLwgBHAAksjAEVEghMxsHNywiCgY2LgimqTKrgCQggDSUPIIUADWECBwAGZQBMIWCSj8mBChkISCCOgB5BCoUI4Vwuquqh5hjJHRAGq8vugQSGms7Fw8-EIzYgRDI2MTjg1Nre1QAPT3UHgAtG-vb1DDo+MLXn6BYL9LY7DL7bKJPI4VCFMTJVbrTbbAFBBAhMLwmLxRKYwjqWZ-aDsFBiFFAsLbQSeCAtKAACQIAFkADIMBCYAAWuDJaPo1NpDJZABF5AA3HnAGauIA)

But notice if I use `Promise` instead of `InstanceType`, the first issues goes away, although `number` is still needed in `ArrayValues`:

```ts
type Constructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

type ElementTypeArrayToInstArray<T extends Constructor[]> = {
	[K in keyof T]: Promise<T[K]> // <-------- no number
}
type ArrayValues<T extends any[]> = T[number & keyof T] // <-------- number still needed
type ElementTypes<T extends Constructor[]> = ArrayValues<ElementTypeArrayToInstArray<T>>

type test = ElementTypes<[typeof HTMLAnchorElement, typeof HTMLDivElement]>
```

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6oDwAqUAvFHAEYBWE2ANFAIJQQAewECAJklAIYIgA2gF0SfASPoBlYL2ABLTGIDeAXwB8YgBQIIAdyhaAdCd4AuRgEoSmgtYBkUGXMUAoV6EhQAogBsIALYcwATgEAyoqLwgBHAAksjAEVEghMxsHNywiCgY2LgimqTKrgCQggDSUPIIUADWECBwAGZQBMIWAAqocAHySBCElcKaAPSjUHgAtDOzM1AIcAvoAeQQqK6qHmGMkdEAary+6BBIaazsXDz8QiNiBIIIK2uoUI4NTa3tUOOTc3PLVbrKAoeS+XwLCAQTjQ7ZePyBYKhSBnIgXTI8eCJPI4VCFMTJA5HE5nBFBBAhMKEmLxRLUwjqdTuTzQdgoMRkpFhM6CFktKAACQIAFkADIMBCYAAWuE5FPofNaQrFABF5AA3OXAEauIA)

### 🙁 Actual behavior

Does not work like #26063

### 🙂 Expected behavior

Works like #26063

---

There seems to be some sort of special casing going on.

**Comment by RyanCavanaugh** [comment]
It seems like you figured this out correctly. What part of this is supposed to be the defect?

**Comment by brabeji** [comment]
@RyanCavanaugh @trusktr Hello guys. Just ran into this. From what I’m able to say, mapping tuples doesn’t seem to work as in #26063 

[Minimal reproduction playground](https://www.typescriptlang.org/play?ts=4.3.4#code/C4TwDgpgBAQg9gDwDwBUoQcCA7AJgZyn2ACcBLbAcwD4oBeKAbygDcBDAGwFcIAuKNAF8AsAChQkKAFkQKLmA7QGAbWLkqAGigA6XWoqVlAXSNiJ0AIL0mYqHajKA0lApQA1hBBwAZtNnzFI354ZBk5BQgnI2oxEVEgA)

What am I missing?

**Comment by RyanCavanaugh** [comment]
`keyof MyTuple` includes things like `"indexOf"` because tuples are arrays. This is the expected behavior.

You can write this instead:
```ts
type Box<T extends string> = { value: T }
type MyTuple = [string, ...string[]]
type A = {
    [K in (keyof MyTuple) & number]: Box<MyTuple[K]>
}
```

**Comment by dtwiers** [comment]
Is this the same issue that is catching me here?

```typescript
type SomeAbstractedType = string;

type SpecialObject<T extends SomeAbstractedType> = { foo: T }

type SpecialObjectsOf<Ts extends SomeAbstractedType[]> = {
    [K in keyof Ts]: SpecialObject<Ts[K]>;
//                                 ^^^^^
// Type 'Ts[K]' does not satisfy the constraint 'string'.
//   Type 'Ts[keyof Ts]' is not assignable to type 'string'.
//     Type 'Ts[string] | Ts[number] | Ts[symbol]' is not assignable to type 'string'.
//       Type 'Ts[string]' is not assignable to type 'string'.(2344)
};
```
If so, what it boils down to, seems to be this: I don't think we can do type constraints on mapped tuples...due to (imo) unintuitive behavior of `keyof MyTupleType` not necessarily being `number`. I'm not sure when I'd ever want `keyof MyTupleType` to include `'length` or `'indexOf'` - perhaps if I want that I would put `keyof Array.prototype` or something.

Relevant TS playground:
https://www.typescriptlang.org/play?ts=4.6.4#code/C4TwDgpgBAyg9gWwgQQEYGdgCcCGBjYCAEwBVxoBeKTLASwDsBzAbgFgAoD0SWSPWnABsA8qgBWEAgB4SUCAA9C9IuliIUGbPkKlyAPihUA3lABmcOAC4osgL4cu5XpIEjxk4OmGmZqhUpU1JDQabWIySABtAF0DYw4oRKhIgGkoBigAawgQOFMbdGjrGD5XUQlpEnRU2LZ2WzqOIklBHCxoPDh6TDMLa0iAcnM4AYAaKAHUNoHoxvZmvFb2qE7u4CgprGLSoXKPLx9uCDzeuD0gA

**Comment by RyanCavanaugh** [comment]
@dtwiers looks like that's been fixed; there's no error in TypeScript 4.7

Closing since there doesn't seem to be anything actionable left
